### PR TITLE
[SPARK-32243][SQL]HiveSessionCatalog call super.makeFunctionExpression should throw earlier when got Spark UDAF Invalid arguments number error

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
When we create a UDAF function use class extended `UserDefinedAggregeteFunction`,  when we call the function,  in support hive mode, in HiveSessionCatalog, it will call super.makeFunctionExpression, 

but it will catch error  such as the function need 2 parameter and we only give 1, throw exception only show 
```
No handler for UDF/UDAF/UDTF xxxxxxxx
```
This is confused for develop , we should show error thrown by super method too, 

For this pr's UT :
Before change, throw Exception like 
```
No handler for UDF/UDAF/UDTF 'org.apache.spark.sql.hive.execution.LongProductSum'; line 1 pos 7
```
After this pr, throw exception
```
Spark UDAF Error: Invalid number of arguments for function longProductSum. Expected: 2; Found: 1;
Hive UDF/UDAF/UDTF Error: No handler for UDF/UDAF/UDTF 'org.apache.spark.sql.hive.execution.LongProductSum'; line 1 pos 7
```

### Why are the changes needed?
Show more detail error message when define UDAF

### Does this PR introduce _any_ user-facing change?
People will see more detail error message when use spark sql's UDAF  in hive support Mode

### How was this patch tested?
Added UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
cc @dongjoon-hyun @maropu

**Comment by AngersZhuuuu** [comment]
cc @dongjoon-hyun @maropu

**Comment by AngersZhuuuu** [comment]
retest this please

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala [hunk 1]
```diff
@@ -33,101 +33,6 @@ import org.apache.spark.sql.test.SQLTestUtils
 import org.apache.spark.sql.types._
 import org.apache.spark.unsafe.UnsafeAlignedOffset
 
-
-class ScalaAggregateFunction(schema: StructType) extends UserDefinedAggregateFunction {
```
Comment (HyukjinKwon) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala:37]: @AngersZhuuuu why do we need to move classes? Let's separate refactoring and the fixes when possible.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 2]
```diff
@@ -69,49 +69,56 @@ private[sql] class HiveSessionCatalog(
...
-            udfExpr = Some(HiveUDAFFunction(
-              name,
-              new HiveFunctionWrapper(clazz.getName),
-              input,
-              isUDAFBridgeRequired = true))
-            udfExpr.get.dataType // Force it to check input data types.
-          } else if (classOf[GenericUDTF].isAssignableFrom(clazz)) {
-            udfExpr = Some(HiveGenericUDTF(name, new HiveFunctionWrapper(clazz.getName), input))
-            udfExpr.get.asInstanceOf[HiveGenericUDTF].elementSchema // Force it to check data types.
+      Try(super.makeFunctionExpression(name, clazz, input)) match {
```
Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:72]: @AngersZhuuuu Seems we don't need to change `getOrElse` to `match`.

Comment (maropu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:72]: +1

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDAFSuite.scala [hunk 3]
```diff
@@ -161,6 +161,21 @@ class HiveUDAFSuite extends QueryTest
       checkAnswer(sql("select histogram_numeric(a,2) from abc where a=3"), Row(null))
     }
   }
+
+  test("Hive mode use spark udaf should show error") {
+    val functionName = "longProductSum"
+    val functionClass = "org.apache.spark.sql.hive.execution.LongProductSum"
+    withUserDefinedFunction(functionName -> true) {
+      sql(s"CREATE TEMPORARY FUNCTION $functionName AS '$functionClass'")
+      val e1 = intercept[AnalysisException] {
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDAFSuite.scala:170]: nit: `e1` -> `e`

**Comment by maropu** [comment]
Please describe exception messages before/after this PR in the description.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:72]: > @AngersZhuuuu Seems we don't need to change `getOrElse` to `match`.

Confused, with `getOrElse` how can I get the exception

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
> Please describe exception messages before/after this PR in the description.

Done

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDAFSuite.scala (hunk 3 above)
Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDAFSuite.scala]: > nit: `e1` -> `e`

Done

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala]: > @AngersZhuuuu why do we need to move classes? Let's separate refactoring and the fixes when possible.

Make a mistake, change back, thanks

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala (hunk 2 above)
Comment (maropu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:72]: If so, how about using `try` instead of `Try`?

**Comment by maropu** [comment]
> 1. Make Expression failed in SessionCatalog for function name = longProductSum: Invalid number of arguments for function longProductSum. Expected: 2; Found: 1;
> 2. No handler for UDF/UDAF/UDTF 'org.apache.spark.sql.hive.execution.LongProductSum'; line 1 pos 7

We need the number `1` and `2`? How about reformatting it like this?
```
No handler for UDF/UDAF/UDTF 'org.apache.spark.sql.hive.execution.LongProductSum' because; Invalid number of arguments for function longProductSum
```

**Comment by AngersZhuuuu** [comment]
> We need the number `1` and `2`? How about reformatting it like this?
> 
> ```
> No handler for UDF/UDAF/UDTF 'org.apache.spark.sql.hive.execution.LongProductSum' because; Invalid number of arguments for function longProductSum
> ```

For me , I prefer more clear information let user know it's wrong form `super.makeExpressionFunction`. and 
```
No handler for UDF/UDAF/UDTF 'org.apache.spark.sql.hive.execution.LongProductSum
```
don't have relation about 
```
Invalid number of arguments for function longProductSum
```

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala (hunk 2 above)
Comment (HyukjinKwon) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: `try` is the way as in the guideline. But in general I would like to avoid changing the indentation for unrelated codes and make it difficult to see what's the real diff. Such pattern makes it more difficult to backport and revert. Let's avoid this @AngersZhuuuu next time. The actual diff seems just adding more message in the exception.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDAFSuite.scala [hunk 4]
```diff
@@ -161,6 +161,21 @@ class HiveUDAFSuite extends QueryTest
       checkAnswer(sql("select histogram_numeric(a,2) from abc where a=3"), Row(null))
     }
   }
+
+  test("Hive mode use spark udaf should show error") {
```
Comment (HyukjinKwon) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDAFSuite.scala:165]: Let's add a JIRA prefix.

Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDAFSuite.scala:165]: > Let's add a JIRA prefix.

Done

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: > `try` is the way as in the guideline. But in general I would like to avoid changing the indentation for unrelated codes and make it difficult to see what's the real diff. Such pattern makes it more difficult to backport and revert. Let's avoid this @AngersZhuuuu next time. The actual diff seems just adding more message in the exception.

Since I always backport pr, I know the point you mentioned,  I will notice this more carefully. 
Yea, the actual diff is to show more message but since we need to know the exception in return Failure(exception), so we can't use `getOrElse()`

**Comment by maropu** [comment]
> For me , I prefer more clear information let user know it's wrong form super.makeExpressionFunction. and

But, the relationship of `1.` and `2.` looks ambiguous to me. Does the message mean that there are two reasons for the failure?  Could you brush up the message more?

**Comment by AngersZhuuuu** [comment]
> But, the relationship of `1.` and `2.` looks ambiguous to me. Does the message mean that there are two reasons for the failure? Could you brush up the message more?

Yea, when we use Spark SQL in support hive mode, the real SessionCatalog is HiveSessionCatalog, when we create and use UDAF,  it call `makeExpressionFunction`. In the code for UDAf, you can see that it will call `super.makeExpressionFunction` first, so it will check if the UDAF is follow spark's UDAF (class implement UserDefinedAggregateFunction), if failed, it will try  hive's UDAF solution. 

So in my test case, it's follow spark's UDAF definition, but `makeExpressionFunction` failed since when I use it, the argument number is wrong. In origin code, it will only show 
```
No handler for UDF/UDAF/UDTF 'org.apache.spark.sql.hive.execution.LongProductSum'
```
But what we real need know is 
```
Invalid number of arguments for function longProductSum
```

These two error is from different level 1. Spark UDAF 2. Hive UDAF.
`1.` `2.` seems really  ambiguous,  how about `For Spark UDAF: ` & `For Hive UDAF` to make it more clear

**Comment by AngersZhuuuu** [comment]
@maropu How about current, I have updated to pr desc.

**Comment by AngersZhuuuu** [comment]
cc @HyukjinKwon @maropu  Any update?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 5]
```diff
@@ -69,49 +69,59 @@ private[sql] class HiveSessionCatalog(
     // Current thread context classloader may not be the one loaded the class. Need to switch
     // context classloader to initialize instance properly.
     Utils.withContextClassLoader(clazz.getClassLoader) {
-      Try(super.makeFunctionExpression(name, clazz, input)).getOrElse {
```
Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:72]: @AngersZhuuuu, can you get rid of this unrelated diffs?

Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:72]: > @AngersZhuuuu, can you get rid of this unrelated diffs?

See  https://github.com/apache/spark/pull/29054#discussion_r452901062
Don't know how can I get the exception without change identation

Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:72]: > @AngersZhuuuu, can you get rid of this unrelated diffs?

How about current change, it won't change indentation.

**Comment by AngersZhuuuu** [comment]
ping @maropu @HyukjinKwon @cloud-fan Any more advise?

**Comment by AngersZhuuuu** [comment]
retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 6]
```diff
@@ -69,7 +69,11 @@ private[sql] class HiveSessionCatalog(
...
+      val t = Try(super.makeFunctionExpression(name, clazz, input))
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:72]: I think we should fail earlier if `super.makeFunctionExpression` fails due to invalid inputs.

Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:72]: > I think we should fail earlier if `super.makeFunctionExpression` fails due to invalid inputs.

Got you point， if throw exception about Invalid input means it's a Spark UDF, and this place only throw this type exception  when match but failed.

Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:72]: > I think we should fail earlier if `super.makeFunctionExpression` fails due to invalid inputs.

When throw `No handler for UDF/UDAF/UDTF` mean both spark and hive UDF definition are not matched

Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:72]: > I think we should fail earlier if `super.makeFunctionExpression` fails due to invalid inputs.

@cloud-fan How about  current change.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 7]
```diff
@@ -69,7 +69,14 @@ private[sql] class HiveSessionCatalog(
...
+      if (t.isSuccess) {
+        t.asInstanceOf[Success[Expression]].get
+      } else {
+        val exception = t.asInstanceOf[Failure[Expression]].exception
+        if (exception.getMessage.contains("Invalid number of arguments for function")) {
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:77]: can we use a special exception type instead of relying on error message?

**Comment by AngersZhuuuu** [comment]
retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala (hunk 7 above)
Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: > can we use a special exception type instead of relying on error message?

Add a special exception for function invalid argument , this action implement form AnalysisException, finally it will be wrapperd as AnalysisException by method `withPosition()` and throw still as `AnalysisException`.

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan @maropu @HyukjinKwon

File: sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala [hunk 8]
```diff
@@ -54,3 +54,13 @@ class AnalysisException protected[sql] (
     s"$message;$lineAnnotation$positionAnnotation"
   }
 }
+
+/**
+ * Thrown when a query failed for invalid function argument size, usually because query pass
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala:59]: `usually because a SQL function is called with the wrong number of arguments.`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 9]
```diff
@@ -1299,8 +1299,8 @@ class SessionCatalog(
 
       // Check input argument size
       if (e.inputTypes.size != input.size) {
-        throw new AnalysisException(s"Invalid number of arguments for function $name. " +
-          s"Expected: ${e.inputTypes.size}; Found: ${input.size}")
+        throw new InvalidFunctionArgumentException(s"Invalid number of arguments for " +
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1302]: Instead of using a special exception here, maybe it's better to create a `InvalidUDFClassException` for the else branch, as it's not really a plan analysis error.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 10]
```diff
@@ -69,7 +69,14 @@ private[sql] class HiveSessionCatalog(
     // Current thread context classloader may not be the one loaded the class. Need to switch
     // context classloader to initialize instance properly.
     Utils.withContextClassLoader(clazz.getClassLoader) {
-      Try(super.makeFunctionExpression(name, clazz, input)).getOrElse {
+      val t = Try(super.makeFunctionExpression(name, clazz, input))
+      if (t.isSuccess) {
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:73]: I'd prefer using native `try catch` instead of writing `if else` for `Try`.

Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:73]: > I'd prefer using native `try catch` instead of writing `if else` for `Try`.

To keep indentation so I change use `try catch` ....  
 https://github.com/apache/spark/pull/29054#discussion_r452812103

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 9 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1302]: > Instead of using a special exception here, maybe it's better to create a `InvalidUDFClassException` for the else branch, as it's not really a plan analysis error.

Raise a new pr or just fix this in this pr

File: sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala (hunk 8 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala:59]: > usually because a SQL function is called with the wrong number of arguments.

Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 9 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1302]: > Instead of using a special exception here, maybe it's better to create a `InvalidUDFClassException` for the else branch, as it's not really a plan analysis error.

So you mean add a exception of 
```
    } else {
      throw new AnalysisException(s"No handler for UDAF '${clazz.getCanonicalName}'. " +
        s"Use sparkSession.udf.register(...) instead.")
    }
```

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1302]: yea

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1302]: and `InvalidUDFClassException` doesn't need to be a public class.

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1302]: > and `InvalidUDFClassException` doesn't need to be a public class.

emmmm， see latest change,  maybe I miss understand what you really want to do.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala [hunk 11]
```diff
@@ -54,3 +54,22 @@ class AnalysisException protected[sql] (
...
+class InvalidFunctionArgumentException protected[sql](message: String)
+  extends AnalysisException(message, None, None, None, None) {
+}
+
+/**
+ * Thrown when a query failed for invalid function class, usually because a SQL
+ * function's class does not follow the rules of the UDF/UDAF class definition.
+ * @param message
+ */
+class InvalidUDFClassException protected[sql](message: String)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala:73]: We can put it under the same package of `SessionCatalog`. This file contains public (user-facing) classes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala [hunk 12]
```diff
@@ -54,3 +54,22 @@ class AnalysisException protected[sql] (
...
   }
 }
+
+/**
+ * Thrown when a query failed for invalid function argument size, usually because
+ * a SQL function is called with the wrong number of arguments.
+ *
+ * @Since 3.1.0
+ */
+class InvalidFunctionArgumentException protected[sql](message: String)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala:64]: This is not needed now.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 13]
```diff
@@ -69,49 +69,56 @@ private[sql] class HiveSessionCatalog(
...
-          } else if (classOf[GenericUDTF].isAssignableFrom(clazz)) {
-            udfExpr = Some(HiveGenericUDTF(name, new HiveFunctionWrapper(clazz.getName), input))
-            udfExpr.get.asInstanceOf[HiveGenericUDTF].elementSchema // Force it to check data types.
+      try {
+        super.makeFunctionExpression(name, clazz, input)
+      } catch {
+        case e: InvalidFunctionArgumentException =>
+          throw e
+        case _: Throwable =>
+          var udfExpr: Option[Expression] = None
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:78]: We can move the code below to a new method.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala (hunk 11 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala:73]: > We can put it under the same package of `SessionCatalog`. This file contains public (user-facing) classes.

Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala (hunk 12 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala:64]: > This is not needed now.

Yea, got your point now

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala (hunk 13 above)
Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:78]: > We can move the code below to a new method.

Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InvalidUDFClassException.scala [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+
+package org.apache.spark.sql.catalyst.catalog
+
+import org.apache.spark.sql.AnalysisException
+
+/**
+ * Thrown when a query failed for invalid function class, usually because a SQL
+ * function's class does not follow the rules of the UDF/UDAF/UDTF class definition.
+ *
+ * @Since 3.1.0
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InvalidUDFClassException.scala:26]: this is not needed for non-public class.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 15]
```diff
@@ -1299,12 +1299,12 @@ class SessionCatalog(
 
       // Check input argument size
       if (e.inputTypes.size != input.size) {
-        throw new AnalysisException(s"Invalid number of arguments for function $name. " +
-          s"Expected: ${e.inputTypes.size}; Found: ${input.size}")
+        throw new AnalysisException(s"Invalid number of arguments for " +
+          s"function $name. Expected: ${e.inputTypes.size}; Found: ${input.size}")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1303]: unnecessary change

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 16]
```diff
@@ -69,49 +119,18 @@ private[sql] class HiveSessionCatalog(
...
-          throw new AnalysisException(s"No handler for UDF/UDAF/UDTF '${clazz.getCanonicalName}'")
-        }
+      try {
+        super.makeFunctionExpression(name, clazz, input)
+      } catch {
+        case _: InvalidUDFClassException =>
+          // If `super.makeFunctionExpression` throw `InvalidUDFClassException`, we construct
+          // Hive UDF/UDAF/UDTF with function definition.
+          makeHiveFunctionExpression(name, clazz, input)
+        case e: AnalysisException =>
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:129]: nit: `case e => throw e`

The only exception is `InvalidUDFClassException`, where we need to try with Hive UDF class. For other exceptions, just re-throw.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 17]
```diff
@@ -57,6 +57,56 @@ private[sql] class HiveSessionCatalog(
       parser,
       functionResourceLoader) {
 
+  def makeHiveFunctionExpression(
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:60]: private

Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:60]: > private

Done

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala (hunk 16 above)
Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:129]: > nit: `case e => throw e`
> 
> The only exception is `InvalidUDFClassException`, where we need to try with Hive UDF class. For other exceptions, just re-throw.

Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 15 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1303]: > unnecessary change

Done

**Review by cloud-fan — APPROVED** [review]

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InvalidUDFClassException.scala [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+
+package org.apache.spark.sql.catalyst.catalog
+
+import org.apache.spark.sql.AnalysisException
+
+/**
+ * Thrown when a query failed for invalid function class, usually because a SQL
+ * function's class does not follow the rules of the UDF/UDAF/UDTF class definition.
+ */
+class InvalidUDFClassException private[sql](message: String)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InvalidUDFClassException.scala:26]: We don't need `private[sql]` as it's already in a private package.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by cloud-fan** [comment]
Sorry I misread the test report. The last commit actually failed to compile.

Since the last commit is very minor and the second last commit passed all tests, I'm sending a followup PR to fix compilation instead of reverting it. https://github.com/apache/spark/pull/29955

**Comment by AngersZhuuuu** [comment]
> Sorry I misread the test report. The last commit actually failed to compile.
> 
> Since the last commit is very minor and the second last commit passed all tests, I'm sending a followup PR to fix compilation instead of reverting it. #29955

Yea, I miss this too...
